src/main/scala/nodes/learning/KernelGenerator.scala

```diff
@@ -0,0 +1,107 @@
+package nodes.learning
+
+import workflow.Transformer
+
+import breeze.linalg._
+import breeze.math._
+import breeze.numerics._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.SparkConf
+import org.apache.spark.SparkContext
+import org.apache.spark.SparkContext._
+import pipelines.Logging
+import scala.reflect.ClassTag
+import utils._
+
+
+// Abstract class shared by Kernel functions
+abstract class KernelGenerator extends Transformer[DenseVector[Double], DenseVector[Double]] {
+
+  val trainDotProd:RDD[Double]
+
+  /* Cached KernelBlockBlock (used for training), and corresponding
+   * blockIdx
+   */
+
+  def apply(testMat: RDD[DenseVector[Double]], blockIdxs: Seq[Int], same: Boolean = false): RDD[DenseVector[Double]]
+
+
+}
+
+
+
+class GaussianKernelGenerator(gamma: Double, trainMat: RDD[DenseVector[Double]])
+  extends KernelGenerator
+  with Serializable with Logging {
+
+  val trainDotProd = trainMat.map {x =>
+    (x.t*x).toDouble
+  }
+
+  /**
+   * Return a column block of the kernel test matrix
+   *
+   * The kernel test matrix is a |test| x |train| matrix K_testtrain.
+   * This function returns the columns in blockIdxs of K_testtrain.
+   *
+   * WARNING: Observe the order that test comes first
+   */
+
+  def apply(testMat: RDD[DenseVector[Double]], blockIdxs: Seq[Int], same: Boolean = false): RDD[DenseVector[Double]] = {
```

Comment:
I guess this was just an optimization for the symmetric case. When you want to generate the kernel matrix for your training data. I can remove it and recompute it every time.